imap_processing/ultra/l0/decom_tools.py

```diff
@@ -0,0 +1,255 @@
+    e = value >> mantissa_bit_length  # Extract the exponent
+    m = value & mantissa_mask  # Extract the mantissa
+
+    if e == 0:
+        return m
+    else:
+        return (base_value + m) << (e - 1)
+
+
+def decompress_binary(
+    binary: str, width_bit: int, block: int, array_length: int, mantissa_bit_length: int
+) -> list:
+    """Decompress a binary string.
+
+    Decompress a binary string based on block-width encoding and
+    logarithmic compression.
+
+    This function interprets a binary string where the value of 'width_bits'
+    specifies the width of the following values. Each value is then extracted and
+    subjected to logarithmic decompression.
+
+    Parameters
+    ----------
+    binary : str
+        A binary string containing the compressed data.
+    width_bit : int
+        The bit width that describes the width of data in the block
+    block : int
+        Number of values in each block
+    array_length : int
+        The length of the array to be decompressed.
+    mantissa_bit_length : int
+        The bit length of the mantissa.
+
+    Returns
+    -------
+    list
+        A list of decompressed values.
+
+    Note: Equations from Section 1.2.1.1 Data Compression and Decompression Algorithms
+    in Ultra_algorithm_doc_rev2.pdf.
+    """
+    current_position = 0
+    decompressed_values = []
+
+    while current_position < len(binary):
+        # Read the width of the block
+        width, current_position = read_and_advance(binary, width_bit, current_position)
+        # If width is 0 or None, we don't have enough bits left
+        if width is None or len(decompressed_values) >= array_length:
+            break
+
+        # For each block, read 16 values of the given width
+        for _ in range(block):
+            # Ensure there are enough bits left to read the width
+            if len(binary) - current_position < width:
+                break
+
+            value, current_position = read_and_advance(binary, width, current_position)
+
+            # Log decompression
+            decompressed_values.append(log_decompression(value, mantissa_bit_length))
+
+    return decompressed_values
+
+
+def decompress_image(
+    pixel0: int,
+    binary_data: str,
+    width_bit: int,
+    mantissa_bit_length: int,
+):
+    """Decompresses a binary string representing an image into a matrix of pixel values.
+
+    It starts with an initial pixel value and decompresses the rest of the image using
+    block-wise decompression and logarithmic decompression based on provided bit widths
+    and lengths.
+
+    Parameters
+    ----------
+    pixel0 : int
+        The first, unmodified pixel p0,0.
+    binary_data : str
+        Binary string.
+    width_bit : int
+        The bit width that describes the width of data in the block
+    mantissa_bit_length : int
+        The bit length of the mantissa.
+
+    Returns
+    -------
+    p_decom : List of lists.
+        Decompressed pixel matrix.
```

Comment:
so decompressed pixel matrix is 2D array? may be list return type could be changed to list of 2D pixel matrix?